src/main/java/com/aws/greengrass/clientdevices/auth/iot/Thing.java

```diff
@@ -85,37 +85,43 @@ public Thing clone() {
      * @param certificateId Certificate ID to attach
      */
     public void attachCertificate(String certificateId) {
-        if (attachedCertificateIds.addIfAbsent(certificateId)) {
-            modified = true;
+        if (attachedCertificateIds.containsKey(certificateId)) {
+            return;
         }
+        attachedCertificateIds.put(certificateId, Instant.now());
+        modified = true;
     }
 
     /**
      * Detach a certificate ID.
      * @param certificateId Certificate ID to detach
      */
     public void detachCertificate(String certificateId) {
-        if (attachedCertificateIds.remove(certificateId)) {
+        if (attachedCertificateIds.remove(certificateId) != null) {
             modified = true;
         }
     }
 
     /**
      * Returns copy of attached certificate IDs.
-     * </p>
-     * This list cannot be modified directly. Refer to {@link #attachCertificate(String) attachCertificate} and
+     * <p>
+     * This map should not be modified directly. Refer to {@link #attachCertificate(String) attachCertificate} and
      * {@link #detachCertificate(String) detachCertificate}
-     *
+     * </p>
      * @return Certificate IDs
      */
-    public List<String> getAttachedCertificateIds() {
-        return new ArrayList<>(attachedCertificateIds);
+    public Map<String, Instant> getAttachedCertificateIds() {
+        Map<String, Instant> certIds;
+        synchronized (attachedCertificateIds) {
+            certIds = new HashMap<>(attachedCertificateIds);
```

Comment:
why are we copying?

Reply:
This copy ensures that a clone of `Thing` has its own copy of certIds that will not refer to the other `Thing`'s. ThingRegistry maintains the latest version of locally persisted Thing.

Reply:
Why do we have clones of Things?

Reply:
@jbutler can correct me here, I believe it is so that ThingRegistry can vend Thing copies to anyone else but Registry is the sole entity that can manipulate (attach/detach certificate) or keep track of latest persisted `Thing` version.

Reply:
We wouldn't need to copy it if we followed why I was recommending above and keep the data structure encapsulated.

Reply:
Oh.... we're copying this because of clone?? That's not right. The constructor should copy. Let me remove this